File: extractor/extractor/queue.py
import json

import nats
from nats.aio.msg import Msg
from .content_processor import ContentProcessor


class Queue:
    def __init__(self, conn: nats.NATS, subject: str, content_processor: ContentProcessor):
        self._conn = conn
        self._subject = subject
        self._content_processor = content_processor

    async def subscribe_handler(self, msg: Msg):
        data = msg.data.decode()

        extractor_message = json.loads(data)
        processing_result = await self._content_processor.process_content(str(extractor_message['content']))

        extractor_message['detected_language'] = processing_result['detected_language']
        extractor_message['keywords'] = processing_result['keywords']
        extractor_message['categories'] = processing_result['categories']

        json_data = json.dumps(extractor_message)
        await self._conn.publish('contextual-engine-extractor-response', json_data.encode())

    async def subscribe(self):
        await self._conn.subscribe(self._subject, cb=self.subscribe_handler)

File: api/internal/queue/queue.go
package queue

import (
	"encoding/json"
	"fmt"
	"github.com/nats-io/nats.go"
	"gitlab.com/kushneryk/contextual-enigne/internal/models"
	"go.uber.org/zap"
)

type Queue struct {
	logger *zap.SugaredLogger
	conn   *nats.Conn
}

func New(l *zap.SugaredLogger, c *nats.Conn) *Queue {
	return &Queue{
		logger: l.With("module", "queue"),
		conn:   c,
	}
}

func (q *Queue) PublishExtractorMessage(message models.ExtractorMessage) error {
	body, err := json.Marshal(message)
	if err != nil {
		return fmt.Errorf("marshal.message.error: %w", err)
	}
	q.logger.With("message", string(body)).Info("contextual-engine-message")
	return q.conn.Publish("contextual-engine-extractor", body)
}

func (q *Queue) ReceiveExtractorMessage(cb func(message []byte)) {
	q.conn.Subscribe("contextual-engine-extractor-response", func(m *nats.Msg) {
		cb(m.Data)
		m.Ack()
	})
}
